api/src/main/java/io/opentelemetry/common/Attributes.java

```diff
@@ -0,0 +1,166 @@
+ */
+@Immutable
+public abstract class Attributes<T> {
+  private static final Attributes<Object> EMPTY =
+      new Attributes<Object>() {
+        @Override
+        public void forEach(AttributeConsumer<Object> consumer) {
+          // no-op
+        }
+      };
+
+  /** Iterate over all the key-value pairs of attributes contained by this instance. */
+  public abstract void forEach(AttributeConsumer<T> consumer);
+
+  @SuppressWarnings("unchecked")
+  private static <T> Attributes<T> sortAndFilter(List<Object> data) {
+    // note: this is possibly not the most memory-efficient possible implementation, but it works.
+    TreeMap<String, T> sorter = new TreeMap<>();
+    for (int i = 0; i < data.size(); i++) {
+      String key = (String) data.get(i++);
+      // todo: skip here, favoring the first, or use the TreeMap's built in replacement to favor the
+      // last?
+      if (!sorter.containsKey(key)) {
+        sorter.put(key, (T) data.get(i));
+      }
+    }
+    List<Object> sortedData = new ArrayList<>(sorter.size() * 2);
+    for (Entry<String, T> entry : sorter.entrySet()) {
+      sortedData.add(entry.getKey());
+      sortedData.add(entry.getValue());
+    }
+    return new AutoValue_Attributes_ArrayBackedAttributes<>(sortedData);
+  }
+
+  /** javadoc me. */
+  @SuppressWarnings("unchecked")
+  public static <T> Attributes<T> empty() {
+    return (Attributes<T>) EMPTY;
+  }
+
+  /** javadoc me. */
+  public static <T> Attributes<T> of(String key, T value) {
+    return sortAndFilter(Arrays.asList(key, value));
+  }
+
+  /** javadoc me. */
+  public static <T> Attributes<T> of(String key1, T value1, String key2, T value2) {
+    return sortAndFilter(
+        Arrays.asList(
+            key1, value1,
+            key2, value2));
+  }
+
+  /** javadoc me. */
+  public static <T> Attributes<T> of(
+      String key1, T value1, String key2, T value2, String key3, T value3) {
+    return sortAndFilter(
+        Arrays.asList(
+            key1, value1,
+            key2, value2,
+            key3, value3));
+  }
+
+  /** javadoc me. */
+  public static <T> Attributes<T> of(
+      String key1, T value1, String key2, T value2, String key3, T value3, String key4, T value4) {
+    return sortAndFilter(
+        Arrays.asList(
+            key1, value1,
+            key2, value2,
+            key3, value3,
+            key4, value4));
+  }
+
+  /** javadoc me. */
+  public static <T> Attributes<T> of(
+      String key1,
+      T value1,
+      String key2,
+      T value2,
+      String key3,
+      T value3,
+      String key4,
+      T value4,
+      String key5,
+      T value5) {
+    return sortAndFilter(
+        Arrays.asList(
+            key1, value1,
+            key2, value2,
+            key3, value3,
+            key4, value4,
+            key5, value5));
+  }
+
+  /** javadoc me. */
+  public static <T> Builder<T> newBuilder() {
```

Comment:
@bogdandrutu Any further thoughts on having something other than a Builder for the "lots of attributes" use-case?